lib/matplotlib/axes/_base.py

```diff
@@ -1180,6 +1177,148 @@ def cla(self):
 
         self.stale = True
 
+    """
+    A sublist of Axes children based on their type.
+
+    This exists solely to warn on modification. In the future, the
+    type-specific children sublists will be immutable tuples.
+    """
+    class ArtistList(MutableSequence):
+        """
+        Parameters
+        ----------
+        axes : .axes.Axes
+            The Axes from which this sublist will pull the children Artists.
+        prop_name : str
+            The property name used to access this sublist from the Axes; used
+            to generate deprecation warnings.
+        add_name : str
+            The method name used to add Artists of this sublist's type to the
+            Axes; used to generate deprecation warnings.
+        valid_types : list of type, optional
+            A list of types that determine which children will be returned by
+            this sublist. If specified, then the Artists in the sublist must be
+            instances of any of these types. If unspecified, then any type of
+            Artist is valid (unless limited by *invalid_types*.)
+        invalid_types : tuple, optional
+            A list of types that determine which children will *not* be
+            returned by this sublist. If specified, then Artists in the sublist
+            will never be an instance of these types. Otherwise, no types will
+            be excluded.
+        """
+        def __init__(self, axes, prop_name, add_name,
+                     valid_types=None, invalid_types=None):
+            self._axes = axes
+            self._prop_name = prop_name
+            self._add_name = add_name
+            self._type_check = lambda artist: (
+                (not valid_types or isinstance(artist, valid_types)) and
+                (not invalid_types or not isinstance(artist, invalid_types))
+            )
+
+        def __repr__(self):
+            return f'<Axes.ArtistList of {self._prop_name}>'
+
+        def __len__(self):
+            return sum(self._type_check(artist)
+                       for artist in self._axes._children)
+
+        def __getitem__(self, key):
```

Comment:
We probably should provide our own `__iter__` as well as the one we get from Sequence 

```python
    def __iter__(self):
        i = 0
        try:
            while True:
                v = self[i]
                yield v
                i += 1
        except IndexError:
            return

```

is going to have terrible line-painter issues if someone does `list(ax.lines)` with a large number of lines on the Axes.